sputnikdao2/src/proposals.rs

```diff
@@ -0,0 +1,430 @@
+            self.token.internal_deposit(&receiver_id, amount);
+            PromiseOrValue::Value(())
+        } else if token_id == BASE_TOKEN {
+            Promise::new(receiver_id.clone()).transfer(amount).into()
+        } else {
+            ext_fungible_token::ft_transfer(
+                receiver_id.clone(),
+                U128(amount),
+                None,
+                &token_id,
+                ONE_YOCTO_NEAR,
+                GAS_FOR_FT_TRANSFER,
+            )
+            .into()
+        }
+    }
+
+    /// Executes given proposal and updates the contract's state.
+    fn internal_execute_proposal(&mut self, proposal: &Proposal) -> PromiseOrValue<()> {
+        // Return the bond.
+        Promise::new(proposal.proposer.clone())
+            .transfer(self.data_mut().policy.to_policy().bounty_bond.0);
+        match &proposal.kind {
+            ProposalKind::ChangeConfig { config } => {
+                self.data_mut().config = config.clone();
+                PromiseOrValue::Value(())
+            }
+            ProposalKind::ChangePolicy { policy } => {
+                self.data_mut().policy = policy.clone();
+                PromiseOrValue::Value(())
+            }
+            ProposalKind::AddMemberToRole { member_id, role } => {
+                self.data_mut()
+                    .policy
+                    .to_policy_mut()
+                    .add_member_to_role(role, member_id);
+                PromiseOrValue::Value(())
+            }
+            ProposalKind::RemoveMemberFromRole { member_id, role } => {
+                self.data_mut()
+                    .policy
+                    .to_policy_mut()
+                    .remove_member_from_role(role, member_id);
+                PromiseOrValue::Value(())
+            }
+            ProposalKind::FunctionCall {
+                receiver_id,
+                method_name,
+                args,
+                deposit,
+                gas,
+            } => Promise::new(receiver_id.clone())
+                .function_call(
+                    method_name.clone().into_bytes(),
+                    args.clone().into(),
+                    deposit.0,
+                    *gas,
+                )
+                .into(),
+            ProposalKind::UpgradeSelf { hash } => {
+                upgrade_self(&hash.0);
+                PromiseOrValue::Value(())
+            }
+            ProposalKind::UpgradeRemote {
+                receiver_id,
+                method_name,
+                hash,
+            } => {
+                let code = env::storage_read(&hash.0).expect("ERR_NO_CODE_STAGED");
+                Promise::new(receiver_id.clone())
+                    .function_call(
+                        method_name.clone().into_bytes(),
+                        code,
+                        NO_DEPOSIT,
+                        env::prepaid_gas() - env::used_gas() - GAS_FOR_UPGRADE_REMOTE_PROMISE,
+                    )
+                    .into()
```

Comment:
Wonder if we should replace with this with low level as well. Since passing code around might be expensive. But hopefully it doesn't clone it. But reading it into memory and back is probably expensive.